### PR TITLE
Use bitwise operation when sampling for resource leak detection.

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:

Modulo operations are slow, we can use bitwise operation to detect if resource leak detection must be done while sampling.

Modifications:

- Ensure the interval is a power of two
- Use bitwise operation for sampling
- Add benchmark.

Result:

Faster sampling.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@Scottmitch @nmittler PTAL

File: common/src/main/java/io/netty/util/internal/MathUtil.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    /**
+     * Fast method of finding the next power of 2 greater than or equal to the supplied value.
+     *
+     * If the value is {@code <= 0} then 1 will be returned.
+     *
+     * This method is not suitable for {@link Integer#MIN_VALUE} or numbers greater than 2^30.*
+     * @param value from which to search for next power of 2
+     * @return The next power of 2 or the value itself if it is a power of 2
+     */
+    public static int findNextPositivePowerOfTwo(final int value) {
```
Comment (nmittler) [inline common/src/main/java/io/netty/util/internal/MathUtil.java]: Is this used?

**Comment by nmittler** [comment]
LGTM

**Comment by nmittler** [comment]
@normanmaurer can you add the benchmark results before and after?

File: common/src/main/java/io/netty/util/internal/MathUtil.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        assert value > Integer.MIN_VALUE && value < 0x40000000;
+        return 1 << (32 - Integer.numberOfLeadingZeros(value - 1));
+    }
+
+    /**
+     * Round up the given value to the next power of two or return the given value if it is already power of two.
+     */
+    public static int powerOfTwo(int res) {
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/internal/MathUtil.java]: Can we remove this and just use `findNextPositivePowerOfTwo`?

Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/MathUtil.java]: Yes why not :+1:

File: common/src/main/java/io/netty/util/internal/MathUtil.java (hunk 1 above)
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/MathUtil.java]: See @Scottmitch 's comment :)

File: common/src/main/java/io/netty/util/internal/MathUtil.java (hunk 2 above)
Comment (nmittler) [inline common/src/main/java/io/netty/util/internal/MathUtil.java]: +1 :)

**Comment by normanmaurer** [comment]
@nmittler will run again with both and add the benchmarks here.

**Comment by Scottmitch** [comment]
@normanmaurer - you close by accident? or are you opening another PR?

**Comment by normanmaurer** [comment]
Cherry-picked into 4.0 (577931e8bcf23d860e4fc20f6cf3c07ee354b83b) , 4.1 (a47685b24379f221a5260897fc72297328cd1be9) and master (04d14494911b8a91f91f991a9943bbe5fedfe7b4)

**Comment by normanmaurer** [comment]
@Scottmitch yep.. now it can be closed :)

**Comment by nmittler** [comment]
@normanmaurer did you run the benchmarks? Would be good to see the improvement.

**Comment by normanmaurer** [comment]
@nmittler 

Before:
```
Benchmark                            Mode  Cnt          Score          Error  Units
ResourceLeakDetectorBenchmark.open  thrpt   20  314563023.980 ± 21195477.017  ops/s
```

After:
```
ResourceLeakDetectorBenchmark.open  thrpt   20  331592614.560 ± 11564671.540  ops/s
```

**Comment by nmittler** [comment]
:+1:

**Pull request closed without merging**